### PR TITLE
Respect two chain finalization

### PR DESCRIPTION
**Pull request by bzawisto**
Caller can now make a query for finalized blocks. This PR also addresses presence of `requires_canonical` flag introduced earlier by https://github.com/Zilliqa/zq2/pull/1079. 

Linked issue: https://github.com/Zilliqa/zq2/issues/863

### CODE REVIEW

File: zilliqa/src/node.rs [hunk 1]
```diff
@@ -263,26 +263,65 @@ impl Node {
...
-        match number {
-            BlockNumberOrTag::Latest => self.get_chain_tip(),
-            BlockNumberOrTag::Finalized => self.get_chain_tip().saturating_sub(2),
-            BlockNumberOrTag::Safe => self.get_chain_tip().saturating_sub(2),
+    pub fn resolve_block_number(&self, block_number: BlockNumberOrTag) -> u64 {
+        match block_number {
+            BlockNumberOrTag::Number(n) => n,
             BlockNumberOrTag::Earliest => 0,
+            BlockNumberOrTag::Latest => self.get_chain_tip(),
             BlockNumberOrTag::Pending => self.get_chain_tip(),
```
Comment (DrZoltanFazekas) [inline zilliqa/src/node.rs:271]: `latest` and `pending` have different semantics:
- `latest` = the most recent block that has been proposed and received
- `pending` = the block that we would be propose as next, containing transactions from the mempool

File: zilliqa/src/node.rs [hunk 2]
```diff
@@ -263,26 +263,65 @@ impl Node {
...
+                let Ok(Some(block)) = self.db.get_block_by_view(&view) else {
+                    return 0u64;
+                };
+                block.number()
+            }
+            // From whitepaper: If the proposed block’s view number
+            // is one larger than its QC’s view number and
+            // is larger or equal to the validator’s local
+            // view number, the validator regards the
+            // proposed block as safe (which, in the codebase, is referenced as head_block())
```
Comment (DrZoltanFazekas) [inline zilliqa/src/node.rs:285]: I'm afraid the terminology used in the HotStuff papers I adhere to in the whitepaper caused a misunderstanding. The term `safe` in HotStuff has nothing to do with the `safe` block tag in the eth APIs.

On post-Merge Ethereum, the `safe` block corresponds to a justified checkpoint: https://github.com/ethereum/consensus-specs/blob/v1.3.0/fork_choice/safe-block.md

The term `justified` comes from Tendermint, and it means it has received 2/3 of the validator votes. Note that the term `checkpoint` here has nothing to do with the weak-subjectivity checkpoints Theo is working on. Ethereum FFG (which is based on the Tendermint algorithm) uses the term `checkpoint` to refer to the first slot of an epoch, which is a candidate for justification (1. voting round) and finalization (2. voting round).

Long story short, the `safe` block tag in our consensus refers to the block that the node's highQC points to (highQC means it's the latest = high, and it's a QC = 2/3 validators voted for it).

**Review by rrw-zilliqa — COMMENTED** [review]
Note that `z2/resources/api_macros.tera.md` contains references to what zq2 returns for pending and latest.- need to remember to update these ...

File: zilliqa/src/node.rs [hunk 3]
```diff
@@ -263,26 +263,66 @@ impl Node {
...
+    pub fn resolve_block_number(&self, block_number: BlockNumberOrTag) -> u64 {
+        match block_number {
+            BlockNumberOrTag::Number(n) => n,
             BlockNumberOrTag::Earliest => 0,
+            BlockNumberOrTag::Latest => self.get_chain_tip(),
             BlockNumberOrTag::Pending => self.get_chain_tip(),
-            BlockNumberOrTag::Number(n) => n,
+            BlockNumberOrTag::Finalized => {
+                let Ok(Some(view)) = self.db.get_latest_finalized_view() else {
+                    return 0u64;
```
Comment (rrw-zilliqa) [inline zilliqa/src/node.rs:274]: Let's not use 0u64 as an error indicator? Sooner or later it will escape from this function and we will get code attempting to reference (or return) block 0x0, not 0x .

File: zilliqa/src/consensus.rs [hunk 4]
```diff
@@ -1028,6 +1037,27 @@ impl Consensus {
         Ok(Some((proposal, broadcasted_transactions)))
     }
 
+    pub fn propose_pending_block(&mut self) -> Result<Block> {
+        let (proposal, applied_transactions) = self
```
Comment (rrw-zilliqa) [inline zilliqa/src/consensus.rs:1041]: I can't see this function used anywhere? What does it do and why is it necessary?

File: zilliqa/src/consensus.rs [hunk 5]
```diff
@@ -1028,6 +1037,27 @@ impl Consensus {
...
+
+        // Check if the votes entry is empty - if so - remove it
+        if let Some((signatures, ..)) = self.votes.get(&proposal.hash()) {
+            if signatures.is_empty() {
+                self.votes.remove(&proposal.hash());
+            }
+        }
+
+        // Re-insert all transactions that were executed during block proposal - they were removed from tx_pool and have to be re-added
+        for tx in applied_transactions {
```
Comment (rrw-zilliqa) [inline zilliqa/src/consensus.rs:1053]: I think the way to do this is to have `propose_block_inner()` take a flag which makes it remember to not remove txns from the pool (or indeed do any other mutations), rather than have the code "just remember what propose_block_inner() does and undo it here" - that means there are two sources of truth - propose_block_inner() and this loop, and inevitably they will drift out of alignment with each other, leading to bugs like "if you call eth_getBalance() with block=pending, and you happen to hit the same node in which there are pending transactions, some of them will be dropped". I'm not sure I want to debug that!

**Review by rrw-zilliqa — CHANGES_REQUESTED** [review]
Couple of comments that I think I'm worried enough about to request changes on, but yell if you feel that I ought to be less brutal.

File: zilliqa/src/node.rs (hunk 3 above)
Comment (JamesHinshelwood) [reply zilliqa/src/node.rs:274]: I think you can reasonably and correctly argue that the genesis block (with view 0) is always finalized. In fact I would probably make that argument by changing the signature of `get_latest_finalized_view` to `Result<u64>` and returning 0 if there is nothing in the DB, along with a comment explaining why.

Comment (JamesHinshelwood) [reply zilliqa/src/node.rs:274]: Oh there are two levels of pattern matching here - I would argue handling mapping `None` to `0` is valid, but `Err` to `0` isn't. Make `resolve_block_number` return `Result<u64>` if its fallible.

File: zilliqa/src/node.rs [hunk 6]
```diff
@@ -263,26 +263,66 @@ impl Node {
...
             BlockId::Hash(RpcBlockHash {
                 block_hash,
-                require_canonical: _,
+                require_canonical,
             }) => {
-                // TODO(#863): require_canonical
-                self.consensus.get_block(&block_hash.into())
+                // See https://eips.ethereum.org/EIPS/eip-1898
+                let Ok(Some(block)) = self.consensus.get_block(&block_hash.into()) else {
+                    const RESOURCE_NOT_FOUND: i32 = -32001;
```
Comment (JamesHinshelwood) [inline zilliqa/src/node.rs:302]: I don't think these eth-API level concerns should be happening in `node.rs`. Why not return `None` in this case and map to the error code in the eth API impls?

File: zilliqa/src/consensus.rs (hunk 4 above)
Comment (JamesHinshelwood) [reply zilliqa/src/consensus.rs:1041]: Does `preview_content` not let you view the mempool without `&mut self`?

File: zilliqa/src/node.rs (hunk 3 above)
Comment (theo-zil) [reply zilliqa/src/node.rs]: I may just be being overly pedanting, but I think that `None` makes sense if there are no blocks in the database. Genesis always being finalized sounds correct, but if you don't even know/have genesis, then your "latest finalized block" isn't genesis, it's none.

Comment (JamesHinshelwood) [reply zilliqa/src/node.rs]: Maybe we should document this invariant more clearly, but I'm 99% sure its impossible for `Consensus` to exist with fewer than one block (i.e. genesis always exists), therefore there is always a finalised block.

Comment (theo-zil) [reply zilliqa/src/node.rs]: That's a good point. Then we can remove the optionality from a few database methods. (Probably in another PR, no need to hold up this one)

**Review by theo-zil — APPROVED** [review]
Personally I'm not a big fan of making the pool and state into mutexes - the ideal solution would be to have a mechanism to operate on it non-mutably to select transactions into a block, and then "commit" the changes once we've settled on a block to build on. There's no real need for concurrency, so having a mutex just feels like the wrong solution compared to having a robust way to "preview" changes non-mutably (e.g. by making a temporary clone somewhere?).

However I also appreciate that this PR has been here for a while and I also didn't get around to voicing my concerns until much earlier, so I'd also be happy with just moving faster and getting this in and potentially refactoring it in a follow-up if a nicer solution can be found. Plus I'm sure that if there was an easy way to solve it (like "just clone it") you'd have implemented it already, so again I'm in favour of merging the feature now and then considering whether it's worth the time to find a "pure" but complex solution later.

File: zilliqa/src/cfg.rs [hunk 7]
```diff
@@ -176,11 +176,11 @@ pub struct ConsensusConfig {
...
 
 pub fn consensus_timeout_default() -> Duration {
-    Duration::from_secs(5)
+    Duration::from_secs(1)
 }
 
 pub fn empty_block_timeout_default() -> Duration {
-    Duration::from_millis(1000)
+    Duration::from_millis(500)
 }
```
Comment (JamesHinshelwood) [inline zilliqa/src/cfg.rs:184]: Why did we change these defaults?

File: zilliqa/src/pool.rs [hunk 8]
```diff
@@ -159,6 +161,10 @@ impl TransactionPool {
                 continue;
             }
 
+            if block_gas_limit.checked_sub(txn.tx.gas_limit()).is_none() {
```
Comment (JamesHinshelwood) [inline zilliqa/src/pool.rs:164]:
```suggestion
            if block_gas_limit < txn.tx.gas_limit() {
```

File: zilliqa/src/state.rs [hunk 9]
```diff
@@ -27,7 +27,7 @@ use crate::{
 /// all the keys are hashed and stored in the same sled tree
 pub struct State {
     db: Arc<TrieStorage>,
-    accounts: PatriciaTrie<TrieStorage>,
+    accounts: Arc<Mutex<PatriciaTrie<TrieStorage>>>,
```
Comment (JamesHinshelwood) [inline zilliqa/src/state.rs:30]: This feels wrong, given `PatriciaTrie` is basically just `root_hash: B256` + `Arc<TrieStorage>` + some caches for mutations. It is already cheaply cloneable, so why do we need the extra layer?

Also, where is the `Mutex` used? When do we need multiple people to have mutable access to the same storage?

**Review by JamesHinshelwood — COMMENTED** [review]
Only half a review, but I think I would benefit from some extra explanation about why we need to change some `&mut self` to `&self` everywhere (and thus I guess why we've wrapped stuff in `Mutex`es?)?
